client/state/plugins/installed/selectors.ts

```diff
@@ -0,0 +1,382 @@
+	( state: AppState ) => {
+		if ( isRequestingForAllSites( state ) ) {
+			return {};
+		}
+
+		return getSiteIdsThatHavePlugins( state ).reduce(
+			( plugins: { [ key: string ]: Plugin }, siteId ) => {
+				if ( isRequesting( state, siteId ) ) {
+					return plugins;
+				}
+
+				const installedPlugins = state.plugins.installed.plugins as InstalledPlugins;
+
+				const pluginsForSite = installedPlugins[ siteId ] || [];
+				pluginsForSite.forEach( ( plugin: InstalledPluginData ) => {
+					const sitePluginProperties = [ 'active', 'autoupdate', 'update', 'version' ];
+					const sitePluginInfo = pick( plugin, sitePluginProperties );
+					const otherPluginInfo = omit( plugin, sitePluginProperties );
+
+					plugins[ plugin.slug ] = {
+						...plugins[ plugin.slug ],
+						...otherPluginInfo,
+						sites: {
+							...plugins[ plugin.slug ]?.sites,
+							[ siteId ]: sitePluginInfo,
+						},
+					};
+				} );
+				return plugins;
+			},
+			{}
+		);
+	},
+	( state ) => [
+		isRequestingForAllSites( state ),
+		getSiteIdsThatHavePlugins( state ),
+		...getSiteIdsThatHavePlugins( state ).map( ( siteId ) => isRequesting( state, siteId ) ),
+	]
+);
+
+/**
+ * The plugins here differ from the plugin objects found on state.plugins.installed.plugins in that each plugin
+ * object has data from all the different sites on it, whereas on state.plugins.installed.plugins they only have
+ * the state for the site which index they are under. The objects here are the same as those returned from
+ * getAllPluginsIndexedByPluginSlug, except they are indexed by siteId.
+ */
+export const getAllPluginsIndexedBySiteId = createSelector(
+	( state ) => {
+		const allPluginsIndexedByPluginSlug = getAllPluginsIndexedByPluginSlug( state );
+
+		return Object.values( allPluginsIndexedByPluginSlug ).reduce(
+			(
+				pluginsIndexedBySiteId: { [ siteId: number ]: { [ pluginSlug: string ]: Plugin } },
+				plugin
+			) => {
+				Object.keys( plugin.sites )
+					.map( ( siteId ) => Number( siteId ) )
+					.forEach( ( siteId ) => {
+						pluginsIndexedBySiteId[ siteId ] = {
+							...pluginsIndexedBySiteId[ siteId ],
+							[ plugin.slug ]: plugin,
+						};
+					} );
+
+				return pluginsIndexedBySiteId;
+			},
+			{}
+		);
+	},
+	( state ) => [ getAllPluginsIndexedByPluginSlug( state ), getSiteIdsThatHavePlugins( state ) ]
+);
+
+export const getFilteredAndSortedPlugins = createSelector(
+	( state: AppState, siteIds: number[], pluginFilter?: PluginFilter ) => {
+		const allPluginsIndexedBySiteId = getAllPluginsIndexedBySiteId( state );
+
+		// Properties on the objects in allPluginsIndexedBySiteId will be modified and the
+		// selector memoization always returns the same object, so use cloneDeep to avoid
+		// altering it for everyone.
+		const allPluginsForSites: { [ pluginSlug: string ]: Plugin } = cloneDeep(
+			siteIds
+				.map( ( siteId: number ) => allPluginsIndexedBySiteId[ siteId ] )
+				.filter( ( plugin ) => !! plugin )
```

Comment:
This might as well be:

```suggestion
				.filter( Boolean )
```